bench/wtperf/wtperf_truncate.c

```diff
@@ -0,0 +1,194 @@
+		trunc_cfg->stone_gap = 0;
+	else
+		trunc_cfg->stone_gap =
+		    (end_point - start_point) / trunc_cfg->needed_stones;
+
+	/* If we have enough data allocate some stones. */
+	if (trunc_cfg->stone_gap != 0) {
+		trunc_cfg->expected_total = (end_point - start_point);
+		for (i = 1; i <= trunc_cfg->needed_stones; i++) {
+			truncate_key = calloc(cfg->key_sz, 1);
+			truncate_item = calloc(sizeof(TRUNCATE_QUEUE_ENTRY), 1);
+			if (truncate_item == NULL) {
+				ret = enomem(cfg);
+				goto err;
+			}
+			generate_key(
+			    cfg, truncate_key, trunc_cfg->stone_gap * i);
+			truncate_item->key = truncate_key;
+			truncate_item->diff =
+			    (trunc_cfg->stone_gap * i) - trunc_cfg->last_key;
+			STAILQ_INSERT_TAIL( &cfg->stone_head, truncate_item, q);
+			trunc_cfg->last_key = trunc_cfg->stone_gap * i;
+			trunc_cfg->num_stones++;
+		}
+	}
+	trunc_cfg->stone_gap = final_stone_gap;
+
+err:	cursor->close(cursor);
+	return (ret);
+}
+
+int
+run_truncate(CONFIG *cfg, CONFIG_THREAD *thread,
+    WT_CURSOR *cursor, WT_SESSION *session, int *truncatedp) {
+
+	TRUNCATE_CONFIG *trunc_cfg;
+	TRUNCATE_QUEUE_ENTRY *truncate_item;
+	char *truncate_key;
+	int ret;
+
+	ret = 0;
+	trunc_cfg = &thread->trunc_cfg;
+
+	*truncatedp = 0;
+	/* Update the total inserts */
+	trunc_cfg->total_inserts = sum_insert_ops(cfg);
+	trunc_cfg->expected_total +=
+	    (trunc_cfg->total_inserts - trunc_cfg->last_total_inserts);
+	trunc_cfg->last_total_inserts = trunc_cfg->total_inserts;
+
+	/* We are done if there isn't enough data to trigger a new milestone. */
+	if (trunc_cfg->expected_total <= trunc_cfg->needed_stones)
+		return (0);
+
+	while (trunc_cfg->num_stones < trunc_cfg->needed_stones) {
+		trunc_cfg->last_key += trunc_cfg->stone_gap;
+		truncate_key = calloc(cfg->key_sz, 1);
+		truncate_item = calloc(sizeof(TRUNCATE_QUEUE_ENTRY), 1);
+		if (truncate_item == NULL) {
+			lprintf(cfg, ENOMEM, 0,
+			    "worker: couldn't allocate cursor array");
+			return (ENOMEM);
+		}
+		generate_key(cfg, truncate_key, trunc_cfg->last_key);
+		truncate_item->key = truncate_key;
+		truncate_item->diff = trunc_cfg->stone_gap;
+		STAILQ_INSERT_TAIL(&cfg->stone_head, truncate_item, q);
+		trunc_cfg->num_stones++;
+	}
+
+	/* We are done if there isn't enough data to trigger a truncate. */
+	if (trunc_cfg->num_stones == 0 ||
+	    trunc_cfg->expected_total <= thread->workload->truncate_count)
+		return (0);
+
+	truncate_item = STAILQ_FIRST(&cfg->stone_head);
+	trunc_cfg->num_stones--;
+	STAILQ_REMOVE_HEAD(&cfg->stone_head, q);
+	cursor->set_key(cursor,truncate_item->key);
+	if ((ret = cursor->search(cursor)) != 0) {
+		lprintf(cfg, ret, 0, "Truncate search: failed");
+		return (ret);
+	}
+
+	if ((ret = session->truncate(session, NULL, NULL, cursor, NULL)) != 0) {
+		lprintf(cfg, ret, 0, "Truncate: failed");
+		return (ret);
+	}
+	*truncatedp = 1;
+	trunc_cfg->expected_total -= truncate_item->diff;
+	free(truncate_item->key);
+	free(truncate_item);
+	truncate_item = NULL;
```

Comment:
You should ensure that the cursor has been reset prior to returning from this function if anything has been done with the cursor.